pkg/kv/kvserver/replica_application_result.go

```diff
@@ -366,17 +343,30 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		raftAdmissionMeta:       nil,
 		v2SeenDuringApplication: false,
 	}
-	// If the original proposal had an explicit span, it's an async consensus
-	// proposal and the span would be finished momentarily (when we return to
-	// the caller) if we didn't unlink it here, but we want it to continue
-	// tracking newProposal. We leave it in `origP.ctx` though, since that
-	// context will become unused once the application of this (soft-failed)
-	// proposal concludes, i.e. soon after this method returns, in case there
-	// is anything left to log into it.
+
 	defer func() {
-		if success {
-			origP.sp = nil
+		if !success {
+			return
 		}
+		// If the original proposal had an explicit span, it's an async consensus
+		// proposal and the span would be finished momentarily (when we return to
+		// the caller) if we didn't unlink it here, but we want it to continue
+		// tracking newProposal. We leave it in `origP.ctx` though, since that
+		// context will become unused once the application of this (soft-failed)
+		// proposal concludes, i.e. soon after this method returns, in case there is
+		// anything left to log into it.
+		origP.sp = nil
+		origP.quotaAlloc = nil
+		origP.ec = makeEmptyEndCmds()
+		origP.doneCh = nil
+
+		// If the proposal is synchronous, the client is waiting on the original
+		// proposal. By the time it has to act on the result, a bunch of reproposals
+		// can have happened. Link them all, so that the client can see all
+		// proposals for post-processing.
+		//
+		// TODO(pavelkalinnikov): prove and assert that origP.reproposal was nil.
+		origP.reproposal = newProposal
```

Comment:
This will prevent garbage collection of all the intermediate reproposals. I don't think we hold on to them from anywhere else? Seems like it would be worthwhile to avoid.

I'm guessing the old proposal is discarded after we hit this code path? If so, we really only need the first and last proposal in the chain. We could e.g. have a backreference to the original proposal and update its forward reference to the latest reproposal, or we could have a cancel function that's propagated down to the active reproposal.

Reply:
Reworked this bit. It became a bit more ugly, but the GC benefit seems worth it. Longer term, we should improve the lifecycle.

Reply:
Also, had to drop the invariant check.